File: LeetCode-CN/String/[205]同构字符串.java
//给定两个字符串 s 和 t，判断它们是否是同构的。 
//
// 如果 s 中的字符可以被替换得到 t ，那么这两个字符串是同构的。 
//
// 所有出现的字符都必须用另一个字符替换，同时保留字符的顺序。两个字符不能映射到同一个字符上，但字符可以映射自己本身。 
//
// 示例 1: 
//
// 输入: s = "egg", t = "add"
//输出: true
// 
//
// 示例 2: 
//
// 输入: s = "foo", t = "bar"
//输出: false 
//
// 示例 3: 
//
// 输入: s = "paper", t = "title"
//输出: true 
//
// 说明: 
//你可以假设 s 和 t 具有相同的长度。 
// Related Topics 哈希表


import java.util.HashMap;
import java.util.HashSet;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public boolean isIsomorphic(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        HashMap<Character, Character> map = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            char ss = s.charAt(i);
            char tt = t.charAt(i);
            if (map.containsKey(ss) || map.containsValue(tt)) {
                if (map.containsKey(ss) && map.get(ss) != tt) {   //不能多对一
                    return false;
                }
                if (map.containsValue(tt)) {
                    for (Character ch : map.keySet()) {
                        if (map.get(ch) == tt && ch != ss) {
                            return false;
                        }
                    }
                }
            } else {
                map.put(ss, tt);
            }
//            if (!map.containsKey(ss) && !map.containsKey(tt)) {
//                map.put(ss, tt);
//                map.put(tt, ss);
//            } else {
//                if (map.containsKey(ss) && map.get(ss) != tt) {
//                    return false;
//                }
//                if (map.containsKey(tt) && map.get(tt) != ss) {
//                    return false;
//                }
//            }
        }
        return true;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: LeetCode-Py/String/[205]同构字符串.py
# 给定两个字符串 s 和 t，判断它们是否是同构的。 
# 
#  如果 s 中的字符可以按某种映射关系替换得到 t ，那么这两个字符串是同构的。 
# 
#  每个出现的字符都应当映射到另一个字符，同时不改变字符的顺序。不同字符不能映射到同一个字符上，相同字符只能映射到同一个字符上，字符可以映射到自己本身。 
# 
#  
# 
#  示例 1: 
# 
#  
# 输入：s = "egg", t = "add"
# 输出：true
#  
# 
#  示例 2： 
# 
#  
# 输入：s = "foo", t = "bar"
# 输出：false 
# 
#  示例 3： 
# 
#  
# 输入：s = "paper", t = "title"
# 输出：true 
# 
#  
# 
#  提示： 
# 
#  
#  可以假设 s 和 t 长度相同。 
#  
#  Related Topics 哈希表 字符串 👍 418 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
class Solution(object):
    def isIsomorphic(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        hs = dict()
        i = 0
        while i < len(s):
            k = s[i]
            v = t[i]
            if hs.has_key(k) or v in hs.values():
                if hs.has_key(k) and hs.get(k) != v:
                    return False
                if v in hs.values():
                    for ch in hs.keys():
                        if hs.get(ch) == v and ch != k:
                            return False
                i += 1
            else:
                hs[k] = v
                i += 1
        return True
# leetcode submit region end(Prohibit modification and deletion)
